### DISCUSSION TITLE
More configurable history read functions and recursive history reading

### DISCUSSION THREAD

**Discussion by HarrySky** [FreeOpcUa/opcua-asyncio#558] [Ideas]
Hi, I was trying to use methods related to reading history from server and found some usability issues.

- Right now you can use `Node.read_raw_history` to get history data in some range that ALWAYS includes bounds (`ReturnBounds` is set to True and not configurable)
- Even if `ContinuationPoint` is returned inside `Node.read_raw_history` request from `Node.history_read` - it is unused and not returned from method (you can't continue reading history if there is more than what was returned)

To address those issues I patch `ContinuationPoint` into `Node.history_read` then copy and change `Node.read_raw_history` code so that it reads all history available and sets `details.ReturnBounds` to `False` (see Minimal Reproducible Example below)

Can some of this configuration (`ReturnBounds` and `ContinuationPoint`) be actually added to asyncua code or maybe some `Node.read_history_recursive` method added? I am ready to send a PR, need your advice and opinion

```python
import asyncio
from datetime import datetime
from typing import Optional

from asyncua import Client, ua
from asyncua.common.node import Node
from asyncua.ua.uatypes import NodeId


async def patched_history_read(
    self: Node,
    details: ua.ReadRawModifiedDetails,
    # [PATCHED IN]
    continuation_point: Optional[bytes] = None,
):
    valueid = ua.HistoryReadValueId()
    valueid.NodeId = self.nodeid
    valueid.IndexRange = ''
    # [PATCHED IN]
    valueid.ContinuationPoint = continuation_point
    params = ua.HistoryReadParameters()
    params.HistoryReadDetails = details
    params.TimestampsToReturn = ua.TimestampsToReturn.Both
    params.ReleaseContinuationPoints = False
    params.NodesToRead.append(valueid)
    return (await self.server.history_read(params))[0]


Node.history_read = patched_history_read


async def main():
    opc_client = Client("opc.tcp://127.0.0.1:4840")
    await opc_client.connect()
    node = opc_client.get_node(NodeId(1, 1))

    # COPIED CODE FROM Node.read_raw_history:
    details = ua.ReadRawModifiedDetails()
    details.IsReadModified = False
    details.StartTime = datetime.fromtimestamp(0)
    details.EndTime = datetime.now()
    details.NumValuesPerNode = 0
    # This is not configurable in Node.read_raw_history!
    details.ReturnBounds = False

    # Loop that reads all history available
    continuation_point: Optional[bytes] = None
    while True:
        result = await node.history_read(details, continuation_point)
        result.StatusCode.check()
        continuation_point = result.ContinuationPoint
        history = result.HistoryData.DataValues
        for value in history:
            print(value)

        # No more data available
        if continuation_point is None:
            break


asyncio.run(main())

```

**Comment by oroulet** [comment]
It is very hard to understand from new code, especially when I have not seen/used the hisotry code for years.. But a PR is welcome and then we will better see your proposed changes

**Reply by HarrySky** [reply]
Will send draft PR soon :+1:

**Reply by HarrySky** [reply]
Done! PR #559
